src/policy/ancestor_packages.h

```diff
@@ -0,0 +1,167 @@
+     * grandparent is not an ancestor.
+     * Set in the constructor and then does not change, even if transactions are skipped.
+     * */
+    bool m_ancestor_package_shaped{false};
+
+    struct PackageEntry {
+        enum class State : uint8_t {
+            /** Default value.
+             * If calling LinearizeWithFees(), this tx must have a fee and vsize. */
+            PACKAGE,
+            /** Excluded from FilteredAncestor{Set,FeeAndVsize}.
+             * Should not appear in any transaction's ancestor_subset.
+             * Excluded in LinearizeWithFees().
+             * Set by calling IsIsInMempool() on this tx. */
+            MEMPOOL,
+            /** This and all descendant's FilteredAncestor{Set,FeeAndVsize} will be std::nullopt.
+             * Should not appear in any transaction's ancestor_subset.
+             * Excluded in LinearizeWithFees().
+             * Set by calling IsDanglingWithDescendants() on this tx or its ancestor. */
+            DANGLING,
+        // If a parent is DANGLING, all descendants must also be DANGLING.
+        // If a child is MEMPOOL, all ancestors should also be MEMPOOL.
+        };
+
+        CTransactionRef tx;
+
+        State m_state{State::PACKAGE};
+
+        /** This value starts as std::nullopt when we don't have any fee information yet. It can be
+         * updated by calling LinearizeWithFees() if this entry isn't MEMPOOL or DANGLING. */
+        std::optional<uint32_t> mining_sequence;
+
+        /** Fees of this transaction. Starts as std::nullopt, can be updated using AddFeeAndVsize(). */
+        std::optional<CAmount> fee;
+
+        /** Virtual size of this transaction (depends on policy, calculated externally). Starts as
+         * std::nullopt. Can be updated using AddFeeAndVsize(). */
+        std::optional<int64_t> vsize;
+
+        /** Txids of all PACKAGE ancestors. Populated in AncestorPackage ctor. If an ancestor
+         * becomes MEMPOOL or DANGLING, it is erased from this set. Use FilteredAncestorSet to get
+         * filtered ancestor sets. */
+        std::set<Txid> ancestor_subset;
+
+        /** Txids of all in-package descendant. Populated in AncestorPackage ctor and does not
+         * change within AncestorPackage lifetime. */
+        std::set<Txid> descendant_subset;
+
+        explicit PackageEntry(CTransactionRef tx_in) : tx(tx_in) {}
+
+        // Used to sort the result of Txns, FilteredTxns, and FilteredAncestorSet. Always guarantees
+        // topological sort to the best of our knowledge (see IsSorted()), and puts the most
+        // incentive-compatible subset(s) first if that information is available.
+        //
+        // If ancestor score-based linearization sequence exists for both transactions, the
+        // transaction with the lower sequence number comes first.
+        //    If unavailable or there is a tie, the transaction with fewer in-package ancestors comes first (topological sort).
```

Comment:
I don’t know if it has been discussed before in the review of this PR, though in case of a tie sounds the breaker should not be on the absolute number of in-package ancestor though on their accumulated vsize ? E.g package entry A might have 1 ancestor of 1 kvB and package entry B might have 3 ancestors of 200 vbytes each.

Reply:
The idea here is just to make sure a descendant doesn't come before its child. If A is an ancestor of B, then all of A's ancestors are also B's ancestors. It must be that A.ancestor_count < B.ancestor_count. Using total ancestor vsize would also work, but count is much easier to check.